### ISSUE TITLE
Fail to restart automatically using docker + splash

### ISSUE THREAD

**Issue by cp2587** [scrapinghub/splash#216]
I use the following command to deamonize the process 
'docker run -d -p 8050:8050 --restart=always -v /etc/splash/proxy-profiles:/etc/splash/proxy-profiles scrapinghub/splash:1.5 --maxrss 500' 
but whenever the splash process crash (it gets killed by the system because it takes too much memory even if i have 1GB on the server) docker fails to restart it with the following error:
```python
Traceback (most recent call last):
  File "/app/bin/splash", line 3, in <module>
    from splash.server import main
  File "/usr/local/lib/python2.7/dist-packages/splash/server.py", line 12, in <module>
    from splash.qtutils import init_qt_app
  File "/usr/local/lib/python2.7/dist-packages/splash/qtutils.py", line 12, in <module>
    from PyQt4.QtCore import (QAbstractEventDispatcher, QDateTime, QObject,
ImportError: /usr/lib/python2.7/dist-packages/PyQt4/QtCore.so: cannot read file data: Input/output error
```

**Comment by kmike** [comment]
Hm, I don't know what can cause this error. Searching for 'Input/output error' suggests a hardware failure, but I have not experience with it myself.

One problem which may be related is that Splash can sometimes segfault on shutdown (it seems because C++ destructors order doesn't match Python objects deallocation order), but here the error is at startup, not at shutdown.

**Comment by cp2587** [comment]
I upgraded both my hardware and linux version, the bug doesnt show up againg when i manually send a kill signal to splash. I will update you if it shows up again.
Thanks for your help

**Comment by cp2587** [comment]
Actually i am monitoring the splash server and i see a lot of 'maxrss exceeded 1440 MB, shutting down'. I understand that splash does not cap its memory usage and so you need to ask him to restart whenever it uses more than a set threshold but why is this threshold so rapidily meet ? Also what happen if i don't set a 'maxrss' option ? Does this mean the system will eventually kill the process ? What if i set the --memory option on docker but i don't use maxrss, can i bound the ram to limit and never release it ?

**Comment by kmike** [comment]
> Actually i am monitoring the splash server and i see a lot of 'maxrss exceeded 1440 MB, shutting down'. I understand that splash does not cap its memory usage and so you need to ask him to restart whenever it uses more than a set threshold but why is this threshold so rapidily meet ? 

It depends on what are you doing with Splash. WebKit is memory hungry, and creating screenshots of full web pages can take a lot of RAM because the whole image is created in memory - several hundred MBs per image is not uncommon, and for very large pages it can use more than a GB of RAM. For example, if you're rendering a full screenshot of blog.pinterest.com with width set to 2100, it'll require storing of about 2100*64009 pixels, each pixel is a 4-byte number, so in total it is 500MB+, and Splash may need to make a copy (I'm not 100% sure).

If you're resizing images then using resize=vector should be more memory efficient (it is an argument of render.png and splash:png()).

You may try reducing the amount of web pages processed in parallel using --slots option. There is also a new [_gc](http://splash.readthedocs.org/en/latest/api.html#gc) endpoint; you can try it to reclaim some memory without a restart.

> Also what happen if i don't set a 'maxrss' option ? Does this mean the system will eventually kill the process ? 

Splash will start swapping. What happens next depends on how is your OS configured. It may enevtually use GBs of swap space.

> What if i set the --memory option on docker but i don't use maxrss, can i bound the ram to limit and never release it ?

If I'm not mistaken, --memory option sets swap space equal to the requested RAM, so Splash will first use RAM, then use swap and then I'm not sure what will happen.

**Comment by cp2587** [comment]
But we are use 'execute' endpoint with a very simple lua script that essentially do 
```lua
assert(splash:wait(%s))
return splash:html()
```
I know python has sometimes trouble garbage collecting and releasing memory but it's strange that splash uses so much memory (and seems to not reallow it or not release it ?). using maxrss works but it somewhat cumbersome to have to restart the reactor every few minutes.

**Comment by kmike** [comment]
There is an internal in-memory cache which WebKit uses; maybe _gc endpoint could help you.

**Comment by cp2587** [comment]
Using a lower value for --slots mitigate the problem as the speed at which the memory is consumed is directly linked to this settings however the memory still grows unchecked (exactly like a memory leak) and so the splash server is restarted every 5 minutes. We added a scrapy middleware to sleep and retry whenever the splash server is unresponsive but this is clearly not a satsifying architecture ;(
I had a similar memory issue related in one of my project and i had to use https://docs.python.org/3/library/concurrent.futures.html (there is a backport for python 2.7) like this
```python
with futures.ProcessPoolExecutor(max_workers=1) as executor:
   ...
```
(see http://stackoverflow.com/questions/15455048/releasing-memory-in-python)

Maybe i can help you to solve this problem in the library or i am the only to encounter this problem ?

**Comment by kmike** [comment]
We are solving it by starting multiple Spash instances (one per CPU core - Splash can use only a single CPU core, so it is a good idea anyways) and putting a load balancer in front of them. This has its own issues (I'm not sure I like how the pool works currently when load balancer is used), but this way when one instance is restarting another can continue processing.

It looks like WebKit memory cache is unbound, and it is never cleared, at least in QWebKit 4. So by default there is a "leak" (ever increasing cache). _gc endpoint is an attempt to provide a way to clear these WebKit caches, maybe running a cron script which pings _gc will help with memory usage in your case. I haven't tried it for real workloads though.

There are some settings in http://doc.qt.io/qt-4.8/qwebsettings.html which may help, e.g. http://doc.qt.io/qt-4.8/qwebsettings.html#setObjectCacheCapacities. I tried it briefly and it seems to help a bit, but I haven't had time to check it in details; PRs to e.g. allow configuring setObjectCacheCapacities values at startup are welcome.

**Comment by cp2587** [comment]
Ok i will take a look, thanks a lot. Regarding the '_gc' endpoint, how can i use the latest version with docker ? Doing a docker scapinghub/splash does not pull the master branch is it ?

**Comment by kmike** [comment]
To get _gc endpoint use "scrapinghub/splash:master" Docker image (see  http://splash.readthedocs.org/en/latest/install.html#splash-versions).

**Comment by kmike** [comment]
It could be the case something other than WebKit is leaking. There is [_debug](http://splash.readthedocs.org/en/latest/api.html#debug) endpoint which can return per-class counts of alive objects (for a restricted set of classes); usually it doesn't show anything suspicious, but maybe you'll find something.

**Comment by cp2587** [comment]
yes nothing supicious for me there:
```json
{"leaks": {"LuaRuntime": 5, "SplashQNetworkAccessManager": 1, "ProfilesSplashProxyFactory": 7, "_SplashHttpClient": 4, "LuaRender": 4, "SplashQWebPage": 4, "_LuaThread": 4, "BrowserTab": 4, "_LuaTable": 16, "SplashCookieJar": 4, "QTimer": 3}
```
_gc does help a little and the master version seems to be more stable (now only rebooting every 10 minutes :).
But i guess if the cause is QT, there is nothing to do on your side. I will try threading and load balancing , it does work very well in general case (i am using celery intensively and they also have a unbound memory problem, they solve by closing & reoping threads after a predefined amount of tasks done)

**Comment by dangra** [comment]
can `_gc` be called periodically from within splash itself? a moment to run GC is when Splash memory usage reaches the maxrss threshold set by command line argument.

**Comment by kmike** [comment]
I like the idea - under the pressure try to survive before giving up. There are 2 problems:

1. I think restarts can still be needed, _gc likely won't fix everything. We should decide when to do a cleanup and when to do a restart.
2. Currently what is measured is a peak RSS a process ever used, so the value never decreases.

**Comment by kmike** [comment]
There is an interesting discussion in this ticket, bu I'm not sure what to with it. Closing.
